rio_alpha/alpha.py

```diff
@@ -58,14 +25,29 @@ def _alpha_worker(open_file, window, ij, g_args):
           and a mask of shape (rows, cols) where
           opaque == 0 and transparent == max of dtype
     """
+    src = open_file[0]
 
-    rgb = open_file[0].read(window=window)
-    depth, rows, cols = rgb.shape
-
-    alpha = calc_alpha(rgb,
-                       g_args['ndv'])
+    arr = src.read(window=window)
 
-    rgba = np.append(rgb, alpha[np.newaxis, :, :], axis=0)
+    # Determine Alpha Band
+    if g_args['ndv']:
+        # User-supplied nodata value
+        alpha = mask_exact(arr, g_args['ndv'])
+    else:
+        # Let rasterio decide
+        alpha = src.dataset_mask(window=window)
+
+    # Replace or Add alpha band to input data
+    if arr.shape[0] == 4:
+        # replace alpha band with rasterio dataset_mask
+        # (likely the same but let's not make that assumption)
+        rgba = arr.copy()
+        rgba[3] = alpha
+    elif arr.shape[0] == 3:
+        # stack the dataset_mask to add alpha band
+        rgba = np.append(arr, alpha[np.newaxis, :, :], axis=0)
+    else:
+        raise ValueError("Array must have 3 or 4 bands (RGB or RGBA)")
```

Comment:
This change is the crux of the PR:

If the user supplies a ndv, use the mask_exact technique. Otherwise, infer it based on the heuristics built into Rasterio.
Only add a band if it's RGB to start, if it's RGBA then overwrite band 4 with our new alpha.